avalanche/evaluation/collector.py

```diff
@@ -0,0 +1,173 @@
+"""
+Utilities for metrics collection outside of Avalanche training Templates.
+"""
+
+import json
+
+import numpy as np
+
+from avalanche._annotations import experimental
+
+
+@experimental()
+class MetricCollector:
+    """A simple metric collector object.
+
+    Functionlity includes the ability to store metrics over time and compute
+    aggregated values of them. Serialization is supported via json files.
+
+    You can pass a stream to the `update` method to compute separate metrics
+    for each stream. Metric names will become `{stream.name}/{metric_name}`.
+    When you recover the stream using the `get` method, you need to pass the
+    same stream.
+
+    Example usage:
+
+    .. code-block:: python
+
+        mc = MetricCollector()
+        for exp in train_stream:
+            res = {"Acc": 0.1} # metric dictionary for the current timestep
+            mc.update(res, stream=test_stream)
+        acc_timeline = mc.get("Accuracy", exp_reduce="sample_mean", stream=test_stream)
+
+    """
+
+    def __init__(self):
+        """Init."""
+        self.metrics_res = {}
+
+        self._stream_len = {}  # stream-name -> stream length
+        self._coeffs = {}  # stream-name -> list_of_exp_length
+
+    def _init_stream(self, stream):
+        # we compute the stream length and number of samples in each experience
+        # only the first time we encounter the experience because
+        # iterating over the stream may be expensive
+        self._stream_len[stream.name] = 0
+        coeffs = []
+        for exp in stream:
+            self._stream_len[stream.name] += 1
+            coeffs.append(len(exp.dataset))
+        coeffs = np.array(coeffs)
+        self._coeffs[stream.name] = coeffs / coeffs.sum()
+
+    def update(self, res, *, stream=None):
```

Comment:
Should we add a "step" argument here ? I see that we store all metrics inside this metric_res but if we don't know at what step one metric was logged the previous metrics are kind of hard to use (because we don't know where they come from). I don't know how we could handle that, usually what I do is that I have a global (atomic) step and then if I want to have another kind of step I log it as a metric which is also a function of the atomic step.

In the classical avalanche metrics we always have a "plot_x" argument in the MetricValue object

Reply:
My idea was that with this API the choice of when `update` is called is left to the user, so the `plot_x` is not necessary anymore. I admit that `plot_x` may still be useful in some cases. I'm also not sure how we should do the time aggregation if `plot_x` can skip timesteps.